### PR TITLE
[SPARK-10996][SPARKR] Implement sampleBy() in DataFrameStatFunctions.

### PR DESCRIPTION
**Pull request by sun-rui**

### CODE REVIEW

**Comment by shivaram** [comment]
cc @felixcheung

**Comment by felixcheung** [comment]
It looked in this a bit since I thought I would be working on SPARK-9443

Would it be more R-like to support named list like this:
```
fractions = list(first=0.1, second=0.2)
sampleBy(df, "key", fractions, 0)
```

because then it is already very easy to access a particular fractions
```
> fractions[["first"]]
[1] 0.1
```

I think this is more intuitive as a R user then env()

**Comment by sun-rui** [comment]
@felixcheung, yes, I agree. I will change fractions as a named list.

File: R/pkg/R/DataFrame.R [hunk 1]
```diff
@@ -1795,17 +1795,15 @@ setMethod("fillna",
...
                 stop("value should be an a named list with each name being a column name.")
               }
-
-              # Convert to the named list to an environment to be passed to JVM
-              valueMap <- new.env()
-              for (col in colNames) {
-                # Check each item in the named list is of valid type
-                v <- value[[col]]
+              # Check each item in the named list is of valid type
+              lapply(value, function(v) {
```
Comment (shivaram) [inline R/pkg/R/DataFrame.R]: Why is the change in `fillna` a part of this PR ?

File: R/pkg/R/sparkR.R [hunk 2]
```diff
@@ -163,19 +163,13 @@ sparkR.init <- function(
     sparkHome <- suppressWarnings(normalizePath(sparkHome))
   }
 
-  sparkEnvirMap <- new.env()
-  for (varname in names(sparkEnvir)) {
-    sparkEnvirMap[[varname]] <- sparkEnvir[[varname]]
-  }
+  sparkEnvirMap <- convertNamedListToEnv(sparkEnvir)
```
Comment (felixcheung) [inline R/pkg/R/sparkR.R]: looks like there are other changes in this PR..

**Comment by sun-rui** [comment]
The conversion from a named list to an env to be passed to JVM backend was used in several functions, so I extract it to a common util function that can be reused. Functions are updated to use this util function, but no change to its original logical, just code re-organization.

File: R/pkg/R/utils.R [hunk 3]
```diff
@@ -588,3 +588,13 @@ mergePartitions <- function(rdd, zip) {
 
   PipelinedRDD(rdd, partitionFunc)
 }
+
+# Convert a named list to an environment to be passed to JVM
+convertNamedListToEnv <- function(namedList) {
+  names <- names(namedList)
```
Comment (felixcheung) [inline R/pkg/R/utils.R]: check `length(na.omit(names(namedList))) == length(namedList)`?
```
> a = list(0.1, 0.2)
> length(na.omit(names(a))) == length(a)
[1] FALSE
> a = list(0.1, 0.2)
> names(a) = c("abc")
> length(na.omit(names(a))) == length(a)
[1] FALSE
> a = list(first=0.1, second=0.2)
> length(na.omit(names(a))) == length(a)
[1] TRUE
```

File: R/pkg/R/DataFrame.R (hunk 1 above)
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: nit: this is going through it twice..  perhaps add a FUN param to `convertNamedListToEnv` to allow for a custom validation for each value?

**Comment by felixcheung** [comment]
got it. suggest adding a check, looks good otherwise.

File: R/pkg/R/utils.R (hunk 3 above)
Comment (sun-rui) [inline R/pkg/R/utils.R]: yes, this check makes sense. I will add it.

File: R/pkg/R/stats.R [hunk 4]
```diff
@@ -100,3 +100,36 @@ setMethod("corr",
...
+#' @examples
+#'\dontrun{
+#' df <- jsonFile(sqlContext, "/path/to/file.json")
+#' sample <- sampleBy(df, "key", fractions, 36)
+#' }
+setMethod("sampleBy",
+          signature(x = "DataFrame", col = "character",
+                    fractions = "list", seed = "numeric"),
+          function(x, col, fractions, seed) {
+            fractionsEnv <- convertNamedListToEnv(fractions)
```
Comment (felixcheung) [inline R/pkg/R/stats.R]: also should this support when fractions or seed are omitted/missing?

Comment (sun-rui) [inline R/pkg/R/stats.R]: the signature prevents the required parameters from being missing.

Comment (felixcheung) [inline R/pkg/R/stats.R]: oh right. :)

**Comment by shivaram** [comment]
@sun-rui Change looks pretty good. I didn't notice the refactoring of convertListToEnvironment -- I think thats a good idea. Could you bring this up to date with master ?

**Comment by sun-rui** [comment]
@shivaram, rebased to master.

**Comment by shivaram** [comment]
Thanks @sun-rui LGTM. Merging this.

**Pull request closed without merging**